response.go

```diff
@@ -215,253 +215,378 @@ func visitModelNode(model interface{}, included *map[string]*Node,
 		fieldValue := modelValue.Field(i)
 		fieldType := modelType.Field(i)
 
-		args := strings.Split(tag, annotationSeperator)
+		args := strings.Split(tag, annotationSeparator)
 
 		if len(args) < 1 {
-			er = ErrBadJSONAPIStructTag
-			break
+			return nil, ErrBadJSONAPIStructTag
 		}
 
 		annotation := args[0]
 
 		if (annotation == annotationClientID && len(args) != 1) ||
 			(annotation != annotationClientID && len(args) < 2) {
-			er = ErrBadJSONAPIStructTag
+			return nil, ErrBadJSONAPIStructTag
+		}
+
+		var err error
+
+		switch annotation {
+		case annotationPrimary:
+			node, err = resolveNodeID(node, fieldValue, fieldType)
+
+			if err != nil {
+				return nil, err
+			}
+
+			node.Type = args[1]
+		case annotationClientID:
+			clientID := fieldValue.String()
+			if clientID != "" {
+				node.ClientID = clientID
+			}
+		case annotationAttribute:
+			node = resolveNodeAttribute(node, fieldValue, args)
+		case annotationRelation:
+			node, err = resolveNodeRelation(node, fieldValue, args, model, included, sideload)
+
+			if err != nil {
+				return nil, err
+			}
+		default:
+			return nil, ErrBadJSONAPIStructTag
+		}
+	}
+
+	if linkableModel, isLinkable := model.(Linkable); isLinkable {
+		jl := linkableModel.JSONAPILinks()
+		if er := jl.validate(); er != nil {
+			return nil, er
+		}
+		node.Links = linkableModel.JSONAPILinks()
+	}
+
+	if metableModel, ok := model.(Metable); ok {
+		node.Meta = metableModel.JSONAPIMeta()
+	}
+
+	return node, nil
+}
+
+func resolveNodeID(node *Node, fieldValue reflect.Value, fieldType reflect.StructField) (*Node, error) {
+	v := fieldValue
+
+	// Deal with PTRS
+	var kind reflect.Kind
+	if fieldValue.Kind() == reflect.Ptr {
+		kind = fieldType.Type.Elem().Kind()
+		v = reflect.Indirect(fieldValue)
+	} else {
+		kind = fieldType.Type.Kind()
+	}
+
+	// Handle allowed types
+	switch kind {
+	case reflect.String:
+		node.ID = v.Interface().(string)
+	case reflect.Int:
+		node.ID = strconv.FormatInt(int64(v.Interface().(int)), 10)
+	case reflect.Int8:
+		node.ID = strconv.FormatInt(int64(v.Interface().(int8)), 10)
+	case reflect.Int16:
+		node.ID = strconv.FormatInt(int64(v.Interface().(int16)), 10)
+	case reflect.Int32:
+		node.ID = strconv.FormatInt(int64(v.Interface().(int32)), 10)
+	case reflect.Int64:
+		node.ID = strconv.FormatInt(v.Interface().(int64), 10)
+	case reflect.Uint:
+		node.ID = strconv.FormatUint(uint64(v.Interface().(uint)), 10)
+	case reflect.Uint8:
+		node.ID = strconv.FormatUint(uint64(v.Interface().(uint8)), 10)
+	case reflect.Uint16:
+		node.ID = strconv.FormatUint(uint64(v.Interface().(uint16)), 10)
+	case reflect.Uint32:
+		node.ID = strconv.FormatUint(uint64(v.Interface().(uint32)), 10)
+	case reflect.Uint64:
+		node.ID = strconv.FormatUint(v.Interface().(uint64), 10)
+	case reflect.Struct:
+		if nStr, ok := v.Interface().(sql.NullString); ok {
```

Comment:
Other `sql.Null*` types don't make sense for an ID, so we only consider: `sql.NullString`, `sql.NullInt32` and `sql.NullInt64`.